oximeter/db/src/native/connection.rs

```diff
@@ -0,0 +1,371 @@
+        let (reader, writer) = stream.into_split();
+        let mut reader = FramedRead::new(reader, Decoder);
+        let mut writer = FramedWrite::new(writer, Encoder);
+        let server_info =
+            Self::exchange_hello(&mut reader, &mut writer).await?;
+        Ok(Self {
+            address,
+            server_info,
+            reader,
+            writer,
+            outstanding_query: false,
+        })
+    }
+
+    /// Return the identity of the connected server.
+    pub fn server_info(&self) -> &ServerHello {
+        &self.server_info
+    }
+
+    /// Exchange hello messages with the server.
+    ///
+    /// This is run automatically at the time we connnect to it.
+    async fn exchange_hello(
+        reader: &mut FramedRead<OwnedReadHalf, Decoder>,
+        writer: &mut FramedWrite<OwnedWriteHalf, Encoder>,
+    ) -> Result<ServerHello, Error> {
+        writer.send(ClientPacket::Hello(OXIMETER_HELLO.clone())).await?;
+        let hello = match reader.next().await {
+            Some(Ok(ServerPacket::Hello(hello))) => hello,
+            Some(Ok(packet)) => {
+                probes::unexpected__server__packet!(|| packet.kind());
+                return Err(Error::UnexpectedPacket(packet.kind()));
+            }
+            Some(Err(e)) => return Err(e),
+            None => return Err(Error::Disconnected),
+        };
+        if hello.name != "ClickHouse"
+            || hello.version_major != VERSION_MAJOR
+            || hello.version_minor != VERSION_MINOR
+            || hello.revision != REVISION
+            || hello.version_patch != VERSION_PATCH
+        {
+            return Err(Error::UnrecognizedClickHouseServer {
+                name: hello.name,
+                major_version: hello.version_major,
+                minor_version: hello.version_minor,
+                patch_version: hello.version_patch,
+                revision: hello.revision,
+            });
+        }
+        Ok(hello)
+    }
+
+    /// Send a ping message to the server to check the connection's health.
+    ///
+    /// This is a cheap way to check that the server is alive and the connection
+    /// is still valid. It will await the pong response from the server.
+    pub async fn ping(&mut self) -> Result<(), Error> {
+        self.writer.send(ClientPacket::Ping).await?;
+        match self.reader.next().await {
+            Some(Ok(ServerPacket::Pong)) => Ok(()),
+            Some(Ok(packet)) => {
+                probes::unexpected__server__packet!(|| packet.kind());
+                Err(Error::UnexpectedPacket(packet.kind()))
+            }
+            Some(Err(e)) => Err(e),
+            None => Err(Error::Disconnected),
+        }
+    }
+
+    /// Cancel a running query, if one exists.
+    pub async fn cancel(&mut self) -> Result<(), Error> {
+        if self.outstanding_query {
+            self.writer.send(ClientPacket::Cancel).await?;
+            // TODO(ben) await EOS packet anyway
```

Comment:
I did want to come back to it, yeah, thanks for reminding me. I'm actually not sure we want this interface. Right now, you can't actually _use_ this to cancel a query. Both `Connection::cancel()` and `Connection::query()` require `&mut self`, so you can only run one to completion, rather than use `cancel()` to interrupt `query()`.

I've written a small test where you can do this, but it requires some ceremony.

- Wrap the `Connection` in an `Arc<Mutex<Connection>>`
- Create a channel of some kind, say a oneshot or `Notify`
- Send a clone of the `Arc` and the receive side of the channel into a spawned task, which basically does `tokio::select!()` between those
- In the side waiting on the channel, use `Connection::cancel()` to actually do the cancellation. The `Connection::query()` side is dropped.
- Return the result from that little task

I've tested that it all works, but it's a bit funky. It's also not crucial to have that right now. So I think I'm going to leave this method _private_ but keep the test. That ensures that we _could_ use it to cancel a query, if we go through that ceremony, but we can wait until we learn more about using it or putting it in a connection pool before deciding how to properly expose a cancellation method.

Let me know if you have any thoughts, and I'll put up a commit with that anyway. If you think it's not worth it, we can back it out before merging.

Reply:
Added this test in 220e928dc

Reply:
I trust your judgement here.